Add Heightmap API

Spottedleaf opened the pull request:
Resolves #1672 
This API is intended to expose useful heightmaps found in the server to API. This exposes all of the live world heightmaps currently in the server. 
If a heightmap becomes impossible to implement, api spec allows the implementation to throw UnsupportedOperationException (far better than returning some weird unexpected value).

Tested via:
https://gist.github.com/Spottedleaf/5d47f67c55a9fb870251ff344bfeb6b3

kashike commented on Spigot-API-Patches/0171-Add-Heightmap-API.patch:
```diff
@@ -0,0 +1,112 @@
++     * Enumeration of different heightmap types maintained by the server. Generally using these maps is much faster
++     * than using an iterative search for a block in a given x, z coordinate.
++     */
++    public static enum HeightMapType {
```
Should be `HeightmapType`.

Spottedleaf committed:
Add Heightmap API

This API is intended to expose useful heightmaps found in the server to API.
This exposes all of the live world heightmaps currently in the server.
If a heightmap becomes impossible to implement, api spec allows the
implementation to throw UnsupportedOperationException for unsupported
map types.

Spottedleaf committed:
Add required nullability annotations

aikar commented on Spigot-API-Patches/0182-Add-Heightmap-API.patch:
```diff
@@ -0,0 +1,114 @@
++     * Enumeration of different heightmap types maintained by the server. Generally using these maps is much faster
++     * than using an iterative search for a block in a given x, z coordinate.
++     */
++    public static enum HeightmapType {
```
Lets move this to its own file

aikar commented on Spigot-Server-Patches/0437-Add-Heightmap-API.patch:
```diff
@@ -0,0 +1,45 @@
++    // Paper start - Implement heightmap api
++    @Override
++    public int getHighestBlockYAt(final int x, final int z, final HeightmapType heightmap) throws UnsupportedOperationException {
++        loadChunk(x >> 4, z >> 4);
```
we don't need loadChunk here
replace this method to use `world.a(HeightMap.Type heightmap_type, int i, int j);`
instead and add OBF helper

Spottedleaf replied:
we need loadchunk since the heightmap will return 0 for unloaded chunks

Spottedleaf committed:
Move HeightmapType and add some helper methods to Location

Improve internal implementation by avoiding BlockPosition methods

Spottedleaf committed:
Add documentation to location methods
